lib/Sema/CSApply.cpp

```diff
@@ -7323,6 +7326,59 @@ Expr *ExprRewriter::finishApply(ApplyExpr *apply, Type openedType,
     return finishApply(apply, openedType, locator);
   }
 
+  // Handle call method applications.
+  auto &ctx = cs.getASTContext();
+
+  TupleExpr *arg = dyn_cast<TupleExpr>(apply->getArg());
+  if (auto parenExpr = dyn_cast<ParenExpr>(apply->getArg()))
+    arg = TupleExpr::createImplicit(ctx, parenExpr->getSubExpr(), {});
+
+  // Get resolved call method and verify it.
+  auto loc = locator.withPathElement(ConstraintLocator::ApplyFunction);
+  auto selected = solution.getOverloadChoice(cs.getConstraintLocator(loc));
+  auto choice = selected.choice;
+  auto *callMethod = dyn_cast<FuncDecl>(selected.choice.getDecl());
+  if (callMethod && callMethod->isCallable()) {
+    auto methodType =
+        simplifyType(selected.openedType)->castTo<AnyFunctionType>();
+    auto selfParam = callMethod->getImplicitSelfDecl();
+    // Diagnose `mutating` method call on immutable value.
+    // FIXME(TF-444): This logic for `mutating` method calls incorrectly rejects
+    // IUOs. Performing this ad-hoc check using `originalFn` feels hacky;
+    // rewrite if possible.
+    if (!cs.getType(originalFn)->hasLValueType() && selfParam->isInOut()) {
```

Comment:
Todo: find a better way to handle/diagnose `mutating func call` methods.
Then, re-enable `mutating func call` + IUO test.

Reply:
Constraint system experts: do you have advice on supporting `mutating func callFunction` methods more robustly? Perhaps appropriate constraints can be generated in CSSimplify.cpp?

Currently, function expressions in `ApplyExpr` are always coerced to rvalues, which is problematic for `mutating func callFunction` method calls. The current workaround in CSApply.cpp is ad-hoc and breaks for `mutating func callFunction` + IUO.

[Here's a small reproducer](https://bugs.swift.org/projects/TF/issues/TF-444), with [`-debug-constraint-solver` output](https://gist.github.com/dan-zheng/378403430d08c5a6471957cfc504bdcf):
```swift

struct Mutating {
  var x: Int
  mutating func callFunction() {
    x += 1
  }
}
func testIUO(f: inout Mutating!) {
  f() // not okay
  f.callFunction() // okay
}
```

```
mutating_iuo.swift:8:3: error: cannot use mutating member on immutable value: 'f' is immutable
  f() // not okay
  ^
```

cc @xedin

Reply:
The constraint structure of `ApplicableFunction` should be good enough here, so it's probably just a problem in the solver.  Try doing something like this in  `simplifyApplicableFnConstraint` (where the existing similar call to `getFixedTypeRecursive` is):

```
  Type lvType2 = getFixedTypeRecursive(type2, flags, /*wantRValue=*/false);
  type2 = lvType2->hasLValueType() ? getFixedTypeRecursive(lvType2->getRValueType(), flags, /*wantRValue=*/true) : lvType2;
```

Then you can use lvType2 as the base type when building your constraints for the call to `callAsFunction`.

Reply:
Thanks for the advice!

I found that the blocker for `mutating func callAsFunction` + IUO. It's the [following line](https://github.com/dan-zheng/swift/blob/fd458d32869960744d0dda8390df0d3bc93efa2d/lib/Sema/CSApply.cpp#L7114) in `ExprRewriter::finishApply` in CSApply.cpp:
```cpp
fn = cs.coerceToRValue(fn);
```

`ConstraintSystem::coerceToRValue` calls `TypeChecker::coerceToRValue`, which calls `cs.setType` on expressions and inserts an implicit load expression. Thus, simply caching `auto *originalFn = fn` before the call to `cs.coerceToRValue(fn)` is not effective.

If I comment the line `fn = cs.coerceToRValue(fn)`, then the `mutating func callAsFunction` + IUO example passes, but other lit tests fail.

Any advice/suggestions? If there's some way to clone `fn` or "undo" the effects of `cs.coerceToRValue(fn)`, that should do the trick. I tried creating an `ASTWalker` to undo the effects of `cs.coerceToRValue(fn)`, but I'm not sure how to nicely undo `ConstraintSystem::setType` which expects a non-`nullptr` `Type` argument.

Reply:
Can you look at the function expression and/or recorded overload choices for the call and make the decision about callable vs. dynamicCallable vs. function before doing any processing?

Reply:
> Can you look at the function expression and/or recorded overload choices for the call and make the decision about callable vs. dynamicCallable vs. function before doing any processing?

Yes! I believe that's possible and may circumvent the issue. Experimenting now.